pkg/volume/iscsi/attacher.go

```diff
@@ -159,29 +160,67 @@ func (detacher *iscsiDetacher) UnmountDevice(deviceMountPath string) error {
 func (attacher *iscsiAttacher) volumeSpecToMounter(spec *volume.Spec, host volume.VolumeHost, pod *v1.Pod) (*iscsiDiskMounter, error) {
 	var secret map[string]string
 	var bkportal []string
-	iscsi, readOnly, err := getVolumeSource(spec)
+	readOnly, fsType, err := getISCSIVolumeInfo(spec)
 	if err != nil {
 		return nil, err
 	}
-	// Obtain secret for AttachDisk
-	if iscsi.SecretRef != nil && pod != nil {
-		if secret, err = volumeutil.GetSecretForPod(pod, iscsi.SecretRef.Name, host.GetKubeClient()); err != nil {
-			glog.Errorf("Couldn't get secret from %v/%v", pod.Namespace, iscsi.SecretRef)
+	if pod != nil {
+		chapDiscover, chapSession, err := getISCSICHAPInfo(spec)
+		if err != nil {
 			return nil, err
 		}
+		if chapDiscover || chapSession {
+			secretNamespace, secretName, err := getISCSISecretNamespaceAndName(spec, pod.Namespace)
+			if err != nil {
+				return nil, err
+			}
+
+			// Obtain secret for AttachDisk
+			if len(secretNamespace) > 0 && len(secretName) > 0 {
```

Comment:
it's an error to have a non-nil secretRef and an empty namespace/name, getISCSISecretNamespaceAndName should return an error in that case